### PR TITLE
JP SE-224 Unifies the persistence layer for User and User Progress domains

### PR DESCRIPTION
**Pull request by juanpedromoreno**
Please, @raulraja @dialelo Review when you have a chance. Thanks.

### CODE REVIEW

File: site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+package com.fortysevendeg.exercises.persistence.domain
+
+import cats.data.Xor
+import shared.User
+import shapeless._
+import ops.record._
+
+object UserCreation {
+
+  // TODO: add proper case objects with errors when switching to Postgres
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala]: Can this TODO be removed now?

File: site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  sealed trait CreationError
+
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala]: We should make all root sealed objects in ADTs be abstract classes extending Product with Serializable

File: site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  case object DuplicateName extends CreationError
+
+  case class Request(
+      login:      String,
+      name:       String,
+      githubId:   String,
+      pictureUrl: String,
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala]: What happens when a github user has not set any picture? Do we get back their default github picture?

File: site/server/app/com/fortysevendeg/exercises/persistence/repositories/UserRepository.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  } yield maybeUser
+}
+
+object UserRepository {
+
+  //Queries input:
+  type UpdateParams = (String, String, String, String, Option[String], Long)
+  type InsertParams = (String, String, String, String, String, Option[String])
+
+  implicit def instance(implicit persistence: PersistenceModule): UserRepository = new UserDoobieRepository()
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/persistence/repositories/UserRepository.scala]: No point on the empty parents when invoking the constructor

File: site/server/test/UserQueriesSpec.scala [hunk 5]
```diff
@@ -3,39 +3,25 @@
...
-      login = "47deg",
-      name = "47",
-      githubId = "47deg",
-      pictureUrl = "http://placekitten.com/50/50",
-      githubUrl = "http://github.com/47deg",
-      email = Some("hello@47deg.com")
-    )
-  ))
-}
+  check(Query[Unit, User](Q.all))
```
Comment (raulraja) [inline site/server/test/UserQueriesSpec.scala]: Nice

File: site/server/test/UserStoreSpec.scala [hunk 6]
```diff
@@ -19,31 +18,22 @@ class UserStoreSpec
     with GeneratorDrivenPropertyChecks
     with Matchers
     with DatabaseInstance
+    with ArbitraryInstances
```
Comment (raulraja) [inline site/server/test/UserStoreSpec.scala]: :+1:

File: site/server/app/com/fortysevendeg/exercises/utils/OAuth2.scala [hunk 7]
```diff
@@ -89,7 +88,7 @@ class OAuth2Controller(
           val htmlUrl = (response.json \ "html_url").as[String]
           val email = (response.json \ "email").asOpt[String]
 
-          UserDoobieStore.getOrCreate(
+          UserRepository.instance.getOrCreate(
```
Comment (raulraja) [inline site/server/app/com/fortysevendeg/exercises/utils/OAuth2.scala]: We should summon the instance via implícits so we actually test the real implementation if it ever changes. As it stands this is just getting the actual implementation without DI.

**Comment by raulraja** [comment]
LGTM once my comments are addressed

File: site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala (hunk 3 above)
Comment (juanpedromoreno) [inline site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala]: I don't really now, maybe some default value from the UI is set up. @rafaparadela any thoughts?

Comment (rafaparadela) [inline site/server/app/com/fortysevendeg/exercises/persistence/domain/UserQueries.scala]: I have created a new empty account and I have two news: 

- The good one: GitHub provides an auto-generated avatar that it's not possible to be removed (only replaced).

- The bad one: When an unnamed user is logged the app throws an exception. This PR https://github.com/47deg/scala-exercises/pull/239 solves the problem. @juanpedromoreno @raulraja Could you take a look please?

The result:

**Comment by rafaparadela** [comment]
Please, do not merge until https://github.com/47deg/scala-exercises/pull/239

**Comment by raulraja** [comment]
@juanpedromoreno once you address my comments go ahead and merge this.

**Pull request merged**